pkg/fortify/fpr_to_sarif.go

```diff
@@ -0,0 +1,767 @@
+	Kinds  []string `json:"kinds"`
+}
+
+type Target struct {
+	Id            string        `json:"id"`
+	ToolComponent ToolComponent `json:"toolComponent"`
+}
+
+type ToolComponent struct {
+	Name string `json:"name"`
+	Guid string `json:"guid"`
+}
+
+type SarifRuleProperties struct {
+	Accuracy    string `json:"Accuracy,omitempty"`
+	Impact      string `json:"Impact,omitempty"`
+	Probability string `json:"Probability,omitempty"`
+}
+
+func ConvertFprToSarif(sys System, project *models.Project, projectVersion *models.ProjectVersion, resultFilePath string) error {
+	log.Entry().Debug("Extracting FPR.")
+	_, err := FileUtils.Unzip(resultFilePath, "result/")
+	if err != nil {
+		return err
+	}
+	//File is result/audit.fvdl
+	data, err := ioutil.ReadFile("result/audit.fvdl")
+	if err != nil {
+		return err
+	}
+
+	sarifJSON, err := Parse(sys, project, projectVersion, data)
+	err = ioutil.WriteFile("fortify/result.sarif", sarifJSON, 0700)
+	return err
+}
+
+func Parse(sys System, project *models.Project, projectVersion *models.ProjectVersion, data []byte) ([]byte, error) {
+	//To read XML data, Unmarshal or Decode can be used, here we use Decode to work on the stream
+	reader := bytes.NewReader(data)
+	decoder := xml.NewDecoder(reader)
+
+	var fvdl FVDL
+	decoder.Decode(&fvdl)
+
+	//Now, we handle the sarif
+	var sarif SARIF
+	sarif.Schema = "https://docs.oasis-open.org/sarif/sarif/v2.1.0/cos01/schemas/sarif-schema-2.1.0.json"
+	sarif.Version = "2.1.0"
+	var fortifyRun Runs
+	sarif.Runs = append(sarif.Runs, fortifyRun)
+
+	// Handle results/vulnerabilities
+	for i := 0; i < len(fvdl.Vulnerabilities.Vulnerability); i++ {
+		result := *new(Results)
+		result.RuleID = fvdl.Vulnerabilities.Vulnerability[i].ClassInfo.ClassID
+		result.Level = "none" //TODO
+		//get message
+		for j := 0; j < len(fvdl.Description); j++ {
+			if fvdl.Description[j].ClassID == result.RuleID {
+				result.RuleIndex = j //Seems very abstract
+				rawMessage := fvdl.Description[j].Abstract.Text
+				// Replacement defintions in message
+				for l := 0; l < len(fvdl.Vulnerabilities.Vulnerability[i].AnalysisInfo.ReplacementDefinitions.Def); l++ {
+					rawMessage = strings.ReplaceAll(rawMessage, "Replace key=\""+fvdl.Vulnerabilities.Vulnerability[i].AnalysisInfo.ReplacementDefinitions.Def[l].DefKey+"\"", fvdl.Vulnerabilities.Vulnerability[i].AnalysisInfo.ReplacementDefinitions.Def[l].DefValue)
+				}
+				result.Message = Message{rawMessage}
+				break
+			}
+		}
+
+		//handle properties
+		prop := *new(SarifProperties)
+		prop.InstanceSeverity = fvdl.Vulnerabilities.Vulnerability[i].InstanceInfo.InstanceSeverity
+		prop.Confidence = fvdl.Vulnerabilities.Vulnerability[i].InstanceInfo.Confidence
+		prop.InstanceID = fvdl.Vulnerabilities.Vulnerability[i].InstanceInfo.InstanceID
+		//Use a query to get the audit data
+		// B5C0FEFD-CCB2-4F21-A9D7-87AE600A5885 is "custom rules": handle differently?
+		if result.RuleID == "B5C0FEFD-CCB2-4F21-A9D7-87AE600A5885" {
```

Comment:
This becomes more specific. There can be multiple custom rules. We should somehow generalize this.

Reply:
This ID represents the custom rule that indicates the _presence_ of custom rules as far as I understand, and not this specific custom rule set. If it is not the case, then generalization would require the full list of all IDs that can be identified as "custom rules", or another way of identifying them (the name isn't a safe thing to check for this).